src/test/java/com/wikia/webdriver/common/core/Helios.java

```diff
@@ -97,65 +92,73 @@ public static String getAccessToken(String userName, String password) {
     httpPost.setHeader("X-Wikia-Internal-Request", "0");
 
     try {
-      try {
-        response = httpClient.execute(httpPost);
-      } catch (ConnectTimeoutException e) {
-        PageObjectLogging.log("Timeout when connecting to helios", e, true);
-        response = httpClient.execute(httpPost);
-      }
-
-      HttpEntity entity = response.getEntity();
-      JSONObject responseValue = new JSONObject(EntityUtils.toString(entity));
-
-      EntityUtils.consume(entity);
-
-      token = responseValue.getString("access_token");
+      String token = executeAndRetry(httpPost, extractAccessToken());
       tokenCache.put(userName, token);
-    } catch (JSONException e) {
-      PageObjectLogging.log("JSON EXCEPTION", ExceptionUtils.getStackTrace(e), false);
-      throw new WebDriverException(e);
+      return token;
     } catch (ClientProtocolException e) {
       PageObjectLogging.log("CLIENT PROTOCOL EXCEPTION", ExceptionUtils.getStackTrace(e), false);
       throw new WebDriverException(e);
     } catch (IOException e) {
-      PageObjectLogging.log(IOEXCEPTION_COMMAND,
-          IOEXCEPTION_ERROR_MESSAGE + ExceptionUtils.getStackTrace(e), false);
+      PageObjectLogging
+          .log(IOEXCEPTION_COMMAND, IOEXCEPTION_ERROR_MESSAGE + ExceptionUtils.getStackTrace(e),
+               false);
       throw new WebDriverException(e);
-    } finally {
-      PageObjectLogging.log("LOGIN HEADERS: ",
-        response != null ? response.toString() : null, true);
-      PageObjectLogging.log("LOGIN RESPONSE: ",
-        response != null ? response.getEntity().toString() : null, true);
     }
-    return token;
+  }
+
+  private static List<NameValuePair> prepareLoginParams(String userName, String password) {
+    return Lists.newArrayList(
+        new BasicNameValuePair("grant_type", HeliosConfig.GrantType.PASSWORD.getGrantType()),
+        new BasicNameValuePair("username", userName),
+        new BasicNameValuePair("password", password)
+    );
+  }
+
+  private static ResponseHandler<String> extractAccessToken() {
+    return res -> {
+      HttpEntity entity = res.getEntity();
+      PageObjectLogging.log("LOGIN HEADERS: ", res.toString(), true);
+      PageObjectLogging.log("LOGIN RESPONSE: ", entity.toString(), true);
+      String source = EntityUtils.toString(entity);
+      PageObjectLogging.log("LOGIN RESPONSE RAW: ", source, true);
+      try {
+        JSONObject responseValue = new JSONObject(source);
+        return responseValue.getString("access_token");
+      } catch (JSONException e) {
+        PageObjectLogging.log("JSON EXCEPTION", ExceptionUtils.getStackTrace(e), false);
+        throw new WebDriverException(e);
+      }
+    };
+  }
+
+  private static String executeAndRetry(HttpPost httpPost, ResponseHandler<String> handler)
+      throws IOException {
+    try (CloseableHttpClient httpClient = getDefaultClient()) {
+      try {
+        return httpClient.execute(httpPost, handler);
```

Comment:
This execute method with a handler is much easier to use, as it takes care of closing the connection and consuming the entity - we do not have to do that ourselves.

```
    final HttpEntity entity = response.getEntity();
            try {
                EntityUtils.consume(entity);
            } catch (final Exception t2) {
                // Log this exception. The original exception is more
                // important and will be thrown to the caller.
                this.log.warn("Error consuming content after an exception.", t2);
            }
```